File: week6/sentimental-credit/credit.py
# importing

import cs50
import re


def main():
    # 1- get card number
    num = get_card()
    # get Luhn's sum
    Lu_sum = check_sum(num)

    # check validity and print type
    print("INVALID") if Lu_sum[-1] != "0" else check(num)


def get_card():
    while True:
        try:
            num = cs50.get_string("Number : ")
            # check by re.
            if re.search(r"^[0-9]+$", num):
                return num
            else:
                raise ValueError

        except:
            pass


def check(num):
    if len(num) == 15 and num[0:2] in ["34", "37"]:
        print("AMEX")

    elif len(num) == 16 and num[0:2] in ["51", "52", "53", "54", "55"]:
        print("MASTERCARD")

    elif len(num) == 16 and num[0] == "4":
        print("VISA")

    elif len(num) == 13 and num[0] == "4":
        print("VISA")
    else:
        print("INVALID")

    ...


# this function calculate nummber of digit and Luhn sum
def check_sum(num):
    # get sum of every other digit, starting with the number’s second-to-last digit * 2
    Lsum = 0
    for i in num[-2::-2]:
        if len(str(int(i) * 2)) == 1:
            Lsum += int(i) * 2
        else:
            Lsum += 1 + int(str(int(i) * 2)[-1])

    # add those products’ digits

    # add that sum (13) to the sum of the digits that weren’t multiplied by 2 (starting from the end)
    for i in num[-1::-2]:
        Lsum += int(i)

    return str(Lsum)

    ...


if __name__ == "__main__":
    main()

File: week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int checksum(long number, int *cnt);
void check(long num, int cnt);

int main(void)
{
    int cnt = 0;

    //prompet for input
    long num = get_long("Number : ");

    //calculate checksum
    int val = checksum(num, &cnt);
    printf("%i\n" ,val);
    if (val % 10 == 0)
    {
        check(num, cnt);
    }
    else
    {
        printf("INVALID\n");
    }

}

void check(long num, int cnt)
{
    // sort the type of visa number and validity
    // AMX type
    if ((cnt == 15) && (num / 10000000000000 == 34 || num / 10000000000000 == 37))
    {
        printf("AMEX\n");
    }
    // MASTERCARD type
    else if ((cnt == 16) && (num / 100000000000000 == 51 || num / 100000000000000 == 52 || \
                             num / 100000000000000 == 53 || num / 100000000000000 == 54 || num / 100000000000000 == 55))
    {
        printf("MASTERCARD\n");
    }
    // VISA type
    else if ((cnt == 16) && (num / 1000000000000000 == 4))
    {
        printf("VISA\n");
    }
    else if ((cnt == 13) && (num / 1000000000000 == 4))
    {
        printf("VISA\n");
    }
    // not valid visa
    else
    {
        printf("INVALID\n");
    }


}

// this function for Luhan's algorithm

int checksum(long number, int *cnt)
{

    int digit;
    int sum = 0;
    int unprosum = 0;

    while (number)
    {
        digit = number % 10;
        number /= 10;
        *cnt = *cnt + 1 ;
        if (*cnt % 2 == 0)
        {

            digit = digit * 2;
            if (digit / 10 != 0)
            {
                digit = (digit % 10) + (digit / 10);


            }
            sum = sum + digit;

        }
        else
        {


            unprosum += digit ;
        }

    }
    int chsum = sum + unprosum ;

    return chsum;
}
